### PR TITLE
[5825] XB should consult with the Policy Engine before uploading assets

### PR DESCRIPTION
**Pull request by jvega190**
https://github.com/craftercms/craftercms/issues/5825

### CODE REVIEW

File: ui/app/src/state/actions/preview.ts [hunk 1]
```diff
@@ -38,6 +38,19 @@ interface CommonOperationProps {
   fieldId: string;
 }
 
+interface ElementRecord {
```
Comment (rart) [inline ui/app/src/state/actions/preview.ts]:
```suggestion
// TODO: This is a copy of a Guest package model
interface ElementRecord {
```

File: ui/app/src/assets/guestMessages.ts [hunk 2]
```diff
@@ -149,5 +149,17 @@ export const guestMessages = defineMessages({
...
+    defaultMessage: 'File "{fileName}" doesn\'t comply with project policies and can\'t be uploaded.'
+  },
+  fileNameChangedPolicy: {
+    id: 'desktopAssetUpload.fileNameChangedPolicy',
+    defaultMessage: 'File "{fileName}" was renamed to "{modifiedFileName}" to comply with project policies.'
+  },
+  uploadError: {
+    id: 'desktopAssetUpload.uploadError',
+    defaultMessage: 'Upload error. File was not uploaded.'
   }
```
Comment (rart) [inline ui/app/src/assets/guestMessages.ts]:
```suggestion
    defaultMessage: 'File "{fileName}" doesn\'t comply with project policies and can\'t be uploaded'
  },
  fileNameChangedPolicy: {
    id: 'desktopAssetUpload.fileNameChangedPolicy',
    defaultMessage: 'File "{fileName}" was renamed to "{modifiedFileName}" to comply with project policies'
  },
  uploadError: {
    id: 'desktopAssetUpload.uploadError',
    defaultMessage: 'Upload error. File was not uploaded'
  }
```

File: ui/app/src/state/actions/preview.ts [hunk 3]
```diff
@@ -144,11 +157,18 @@ export const instanceDragEnded = /*#__PURE__*/ createAction('INSTANCE_DRAG_ENDED
 export const navigationRequest = /*#__PURE__*/ createAction('NAVIGATION_REQUEST');
 export const reloadRequest = /*#__PURE__*/ createAction('RELOAD_REQUEST');
 export const desktopAssetDrop = /*#__PURE__*/ createAction('DESKTOP_ASSET_DROP');
-export const desktopAssetUploadComplete = /*#__PURE__*/ createAction<{ record; path: string }>(
+export const desktopAssetUploadComplete = /*#__PURE__*/ createAction<{ record: ElementRecord; path: string }>(
```
Comment (rart) [inline ui/app/src/state/actions/preview.ts]: This couldn't/shouldn't be an element record. An element can't be posted through windows. Double check

File: ui/guest/src/store/epics/root.ts [hunk 4]
```diff
@@ -250,29 +273,108 @@ const epic = combineEpics<GuestStandardAction, GuestStandardAction, GuestState>(
...
-                    })(record.element as HTMLImageElement);
-                    reader.readAsDataURL(file);
-                    return stream$;
+                    const { field } = iceRegistry.getReferentialEntries(record.iceIds[0]);
+                    const {
+                      validations: { allowImageUpload }
+                    } = field;
+
+                    const path =
+                      allowImageUpload && allowImageUpload.value
```
Comment (rart) [inline ui/guest/src/store/epics/root.ts]:
```suggestion
                      allowImageUpload?.value
```

File: ui/guest/src/store/epics/root.ts [hunk 5]
```diff
@@ -250,29 +273,108 @@ const epic = combineEpics<GuestStandardAction, GuestStandardAction, GuestState>(
...
+                        ? processPathMacros({
+                            path: allowImageUpload.value,
+                            objectId: record.modelId
+                          })
+                        : `/static-assets/images/${record.modelId}`;
```
Comment (rart) [inline ui/guest/src/store/epics/root.ts]:
```suggestion
                          // TODO: Support path coming from content type definition
                        : `/static-assets/images/${record.modelId}`;
```

File: ui/guest/src/store/epics/root.ts [hunk 6]
```diff
@@ -250,29 +273,108 @@ const epic = combineEpics<GuestStandardAction, GuestStandardAction, GuestState>(
...
+                                            values: {
+                                              fileName: file.name,
+                                              modifiedFileName: fileName
+                                            }
+                                          })
+                                        );
+                                      }
+                                      return of(
+                                        desktopAssetUploadComplete({
+                                          record: record,
```
Comment (rart) [inline ui/guest/src/store/epics/root.ts]:
```suggestion
                                          record,
```

File: ui/guest/src/store/reducers/root.ts [hunk 7]
```diff
@@ -474,6 +475,15 @@ const reducer = createReducer(initialState, {
     ...state,
     uploading: reversePluckProps(state.uploading, `${record.id}`)
   }),
+  // region desktopAssetUploadFailed
+  [desktopAssetUploadFailed.type]: (
+    state,
+    { payload: { record } }: GuestStandardAction<{ record: ElementRecord }>
+  ) => ({
+    ...state,
+    uploading: reversePluckProps(state.uploading, `${record.id}`)
```
Comment (rart) [inline ui/guest/src/store/reducers/root.ts]:
```suggestion
    uploading: reversePluckProps(state.uploading, String(record.id))
```

**Review by rart — CHANGES_REQUESTED** [review]

File: ui/app/src/state/actions/preview.ts (hunk 3 above)
Comment (jvega190) [reply ui/app/src/state/actions/preview.ts]: Per discussion I moved the desktopAssetUpload actions to guest

**Review by rart — APPROVED** [review]

**Pull request merged**